### DISCUSSION TITLE
Sound Subtraction (modified Noise Reduction)

### DISCUSSION THREAD

**Discussion by GelflingHand** [audacity/audacity#2743] [General]
For example, subtracting the instrumental version of a song from the full version to isolate the vocals. I'm getting decent results by aligning the two tracks, using Get Noise Profile on a tiny segment of the instrumental, then applying Noise Reduction to the matching section of the full song. The problem being that it's unbelievably tedious to do this for any significant duration.

What I'd like is a new Sound Subtraction effect based on the Noise Reduction effect, but instead of a Get Noise Profile button, it requires you to select two tracks, and when activated it selects the smallest possible noise sample (0.05s) from the first track, applies it to the second track, then advances to the next segment, repeating until it gets to the end of the user's selection.

It might be useful to make the target segments slightly longer than the source segments (overlapping each other) to account for less than perfectly matched tracks, eg if there's a bit of an echo in the target track. Maybe fade in and fade out the strength of the noise reduction application on each segment. Ideally there'd be a slider for this.

I've looked into macros and scripting, but it appears we don't have sufficient access to the Noise Reduction functionality via those methods.

**Comment by LWinterberg** [comment]
Try aligning the two tracks, selecting one of the two tracks and then Effects > Invert

**Reply by GelflingHand** [reply]
That only works if the "noise" track is absolutely 100% _identical_ to the noise in the full track, which I've certainly never encountered in real life.

**Comment by Paul-Licameli** [comment]
This sounds to me like a misuse of noise reduction for unintended purposes.

**Reply by GelflingHand** [reply]
It's obviously not what they had in mind when the noise reduction effect was created, no - that's why I'm requesting a new feature. But if the noise reduction algorithm has wider utility than anticipated, why not take advantage of its full capability? It looks to me like a fairly easy thing to implement, and in some cases potentially a lot more effective than existing features like the Vocal Reduction and Isolation plugin.

**Comment by GelflingHand** [comment]
Here's a proof of concept, if anyone who actually knows what they're doing wants to take a look at it:
[SoundSubtraction.h](https://mega.nz/file/2aRyCJTQ#ciw0Ws8CSSBaX7qgyhZzKmpwyA4A6Chn916vQVAYz_Q)
[SoundSubtraction.cpp](https://mega.nz/file/6HwnQIyI#ufRDbVKSujLmgB8N9ztP6702bT8CiRuaXSIdqbeJnJ4)

I've gotten useful results out of it as-is, but there's probably room for improvement since I don't really know C++, the audacity code, or how sound works.

**Reply by Trainmaster9977** [reply]
hello, I've been on the search for something exactly like this, and would really really like to take a look at these files, but they're unavailable for some reason. Mind sending them or hosting them on github? thanks!

**Reply by GelflingHand** [reply]
Hi! Sorry for the late response. Hopefully this attachment works, and I hope the files prove useful to you.
[soundsub.zip](https://github.com/audacity/audacity/files/15211985/soundsub.zip)

**Reply by pszsh** [reply]
This is neat, great work. Not sure why you got so much negativity. I think what you made is really cool and your idea is conceptually intriguing. Thank you.

**Reply by GelflingHand** [reply]
Thanks! It worked adequately for what I needed it for, and it would be nice if others found it useful too.
